pkg/sentry/fsimpl/kernfs/filesystem.go

```diff
@@ -117,24 +117,19 @@ afterSymlink:
 //
 // Postconditions: Caller must call fs.processDeferredDecRefs*.
 func (fs *Filesystem) revalidateChildLocked(ctx context.Context, vfsObj *vfs.VirtualFilesystem, parent *Dentry, name string, child *Dentry) (*Dentry, error) {
-	if child != nil {
-		// Cached dentry exists, revalidate.
-		if !child.inode.Valid(ctx) {
-			delete(parent.children, name)
-			vfsObj.InvalidateDentry(ctx, &child.vfsd)
-			fs.deferDecRef(&child.vfsd) // Reference from Lookup.
-			child = nil
-		}
-	}
 	if child == nil {
 		// Dentry isn't cached; it either doesn't exist or failed revalidation.
 		// Attempt to resolve it via Lookup.
-		c, err := parent.inode.Lookup(ctx, name)
+		c, keep, err := parent.inode.Lookup(ctx, name)
 		if err != nil {
 			return nil, err
 		}
-		// Reference on childVFSD dropped by a corresponding Valid.
-		parent.InsertChildLocked(name, c)
+		if keep {
+			parent.InsertChildLocked(name, c)
+		} else {
+			// FIXME(ayushranjan): two decrefs?
+			fs.deferDecRef(c.VFSDentry())
```

Comment:
It seems we shouldn't decrease the reference count in this case, e.g. runsc will crash when running below command:

```
$ ./bazel-bin/runsc/linux_amd64_pure/runsc --vfs2=true --debug-log=/dev/stderr --network=none --rootless do readlink /proc/self /proc/self
```

```
I0923 19:53:46.819648   78916 main.go:209] Version release-20200914.0-127-gc2a08af80e18
......
panic: Decrementing non-positive ref count 0xc0004cee70, owned by *kernfs.Dentry

goroutine 196 [running]:
panic(0xee9920, 0xc00033e3b0)
        GOROOT/src/runtime/panic.go:1064 +0x46d fp=0xc0001d9758 sp=0xc0001d96a0 pc=0x43438d
gvisor.dev/gvisor/pkg/sentry/fsimpl/kernfs.(*DentryRefs).DecRef(0xc0004cee70, 0xc0001d97d0)
        bazel-out/k8-fastbuild/bin/pkg/sentry/fsimpl/kernfs/dentry_refs.go:110 +0xe2 fp=0xc0001d97c0 sp=0xc0001d9758 pc=0x7a1fb2
gvisor.dev/gvisor/pkg/sentry/fsimpl/kernfs.(*Dentry).DecRef(0xc0004cee70, 0x128ee20, 0xc000288000)
        pkg/sentry/fsimpl/kernfs/kernfs.go:218 +0x69 fp=0xc0001d9800 sp=0xc0001d97c0 pc=0x7ad4f9
gvisor.dev/gvisor/pkg/sentry/vfs.(*Dentry).DecRef(...)
        pkg/sentry/vfs/dentry.go:155
gvisor.dev/gvisor/pkg/sentry/fsimpl/kernfs.(*Filesystem).processDeferredDecRefsLocked(0xc0004d0280, 0x128ee20, 0xc000288000)
        pkg/sentry/fsimpl/kernfs/kernfs.go:129 +0x8c fp=0xc0001d9848 sp=0xc0001d9800 pc=0x7ad2ac
gvisor.dev/gvisor/pkg/sentry/fsimpl/kernfs.(*Filesystem).processDeferredDecRefs(0xc0004d0280, 0x128ee20, 0xc000288000)
        pkg/sentry/fsimpl/kernfs/kernfs.go:121 +0x58 fp=0xc0001d9878 sp=0xc0001d9848 pc=0x7ad1f8
gvisor.dev/gvisor/pkg/sentry/fsimpl/kernfs.(*Filesystem).ReadlinkAt(0xc0004d0280, 0x128ee20, 0xc000288000, 0xc0004be900, 0x0, 0x0, 0x0, 0x0)
        pkg/sentry/fsimpl/kernfs/filesystem.go:534 +0xdf fp=0xc0001d98f8 sp=0xc0001d9878 pc=0x7a7ddf
gvisor.dev/gvisor/pkg/sentry/vfs.(*VirtualFilesystem).ReadlinkAt(0xc000136180, 0x128ee20, 0xc000288000, 0xc00013c0c0, 0xc0001d9a38, 0xaa0001, 0x1, 0xc000480600, 0xc0004bc480)
        pkg/sentry/vfs/vfs.go:450 +0x93 fp=0xc0001d9970 sp=0xc0001d98f8 pc=0x6e1863
gvisor.dev/gvisor/pkg/sentry/syscalls/linux/vfs2.readlinkat(0xc000288000, 0x3ffffff9c, 0x7f6e8864554f, 0x60c030, 0x40, 0x0, 0x0, 0x0, 0x0)
        pkg/sentry/syscalls/linux/vfs2/stat.go:333 +0x1bb fp=0xc0001d9ab0 sp=0xc0001d9970 pc=0xc648bb
gvisor.dev/gvisor/pkg/sentry/syscalls/linux/vfs2.Readlink(0xc000288000, 0x7f6e8864554f, 0x60c030, 0x40, 0x60c030, 0x60c030, 0x0, 0x0, 0x13495, 0x40000002, ...)
        pkg/sentry/syscalls/linux/vfs2/stat.go:240 +0x55 fp=0xc0001d9b08 sp=0xc0001d9ab0 pc=0xc64195
gvisor.dev/gvisor/pkg/sentry/kernel.(*Task).executeSyscall(0xc000288000, 0x59, 0x7f6e8864554f, 0x60c030, 0x40, 0x60c030, 0x60c030, 0x0, 0x0, 0x10ac300, ...)
        pkg/sentry/kernel/task_syscall.go:104 +0x122 fp=0xc0001d9c68 sp=0xc0001d9b08 pc=0x8eee82
gvisor.dev/gvisor/pkg/sentry/kernel.(*Task).doSyscallInvoke(0xc000288000, 0x59, 0x7f6e8864554f, 0x60c030, 0x40, 0x60c030, 0x60c030, 0x0, 0x60c030, 0x0)
        pkg/sentry/kernel/task_syscall.go:239 +0x66 fp=0xc0001d9cf0 sp=0xc0001d9c68 pc=0x8f0066
gvisor.dev/gvisor/pkg/sentry/kernel.(*Task).doSyscallEnter(0xc000288000, 0x59, 0x7f6e8864554f, 0x60c030, 0x40, 0x60c030, 0x60c030, 0x0, 0x125f001, 0x100000000000000)
        pkg/sentry/kernel/task_syscall.go:199 +0x96 fp=0xc0001d9d50 sp=0xc0001d9cf0 pc=0x8efbf6
gvisor.dev/gvisor/pkg/sentry/kernel.(*Task).doSyscall(0xc000288000, 0x2, 0xc000288000)
        pkg/sentry/kernel/task_syscall.go:174 +0x15c fp=0xc0001d9e10 sp=0xc0001d9d50 pc=0x8ef58c
gvisor.dev/gvisor/pkg/sentry/kernel.(*runApp).execute(0x0, 0xc000288000, 0x125f000, 0x0)
        pkg/sentry/kernel/task_run.go:275 +0xc18 fp=0xc0001d9f60 sp=0xc0001d9e10 pc=0x8e3f48
gvisor.dev/gvisor/pkg/sentry/kernel.(*Task).run(0xc000288000, 0x1)
        pkg/sentry/kernel/task_run.go:93 +0x18b fp=0xc0001d9fd0 sp=0xc0001d9f60 pc=0x8e2c5b
runtime.goexit()
        src/runtime/asm_amd64.s:1373 +0x1 fp=0xc0001d9fd8 sp=0xc0001d9fd0 pc=0x467da1
created by gvisor.dev/gvisor/pkg/sentry/kernel.(*Task).Start
        pkg/sentry/kernel/task_start.go:318 +0xfe
```

Reply:
Hmm decrefing does seem like the right thing to do because Lookup should return a dentry with an extra ref that we own.

Reply:
Anyways this fix does not work because we are forced to add something to the dentry tree if an FD is opened on it.